File: solutions/swift/custom-set/1/Sources/CustomSet/CustomSet.swift
struct CustomSet: Equatable {
  var val: [Int]
  var isEmpty: Bool { val.isEmpty }
  init(_ val: [Int]) { self.val = val }

  func union(_ other: CustomSet) -> CustomSet {
    val.reduce(into: CustomSet(other.val)) { $0.add($1) }
  }
  func difference(_ other: CustomSet) -> CustomSet {
    CustomSet(val.filter { !other.val.contains($0) })
  }
  func intersection(_ other: CustomSet) -> CustomSet {
    CustomSet(val.filter { other.val.contains($0) })
  }
  mutating func add(_ el: Int) {
    guard !val.contains(el) else { return }
    self.val.append(el)
  }
  func isDisjoint(with other: CustomSet) -> Bool { val.allSatisfy { !other.val.contains($0) } }
  func isSubset(of other: CustomSet) -> Bool { val.allSatisfy { other.val.contains($0) } }
  func contains(_ el: Int) -> Bool { val.contains(el) }
  static func == (lhs: CustomSet, rhs: CustomSet) -> Bool { lhs.val.sorted() == rhs.val.sorted() }
}

File: solutions/lua/custom-set/1/custom-set.lua
local Set
Set = function(...)
	local s = { d = {} }
	for _, k in ipairs({ ... } or {}) do
		s.d[k] = true
	end
	function s.add(self, x) self.d[x] = true end
	function s.equals(self, other) return #self.d == #other.d and self:is_subset(other) end
	function s.is_empty(self) return #self.d == 0 end
	function s.contains(self, x) return self.d[x] ~= nil end
	function s.is_subset(self, other)
		for k in pairs(self.d) do
			if other.d[k] == nil then return false end
		end
		return true
	end
	function s.is_disjoint(self, other)
		for k in pairs(self.d) do
			if other.d[k] ~= nil then return false end
		end
		return true
	end
	function s.intersection(self, other)
		local retval = Set()
		for k in pairs(self.d) do
			if other.d[k] ~= nil then retval:add(k) end
		end
		return retval
	end
	function s.difference(self, other)
		local retval = Set()
		for k in pairs(self.d) do
			if other.d[k] == nil then retval:add(k) end
		end
		return retval
	end
	function s.union(self, other)
		local retval = self
		for k in pairs(other.d) do
			retval:add(k)
		end
		return retval
	end
	return s
end
return Set
